drivers/stepper/stepper_motion_controllers/zephyr_stepper_motion_controller.c

```diff
@@ -0,0 +1,405 @@
+	const struct stepper_motion_control_config *config = dev->config;
+	struct stepper_motion_control_data *data = dev->data;
+
+	K_SPINLOCK(&data->lock) {
+		data->run_mode = STEPPER_RUN_MODE_VELOCITY;
+		data->direction = direction;
+		config->timing_source->update(&config->timing_config, &data->timing_data,
+					      data->timing_data.microstep_interval_ns);
+		stepper_set_direction(config->stepper, direction);
+		config->timing_source->start(&config->timing_config, &data->timing_data);
+	}
+	return 0;
+}
+
+static int z_stepper_motion_control_stop(const struct device *dev)
+{
+	const struct stepper_motion_control_config *config = dev->config;
+	struct stepper_motion_control_data *data = dev->data;
+	int ret;
+
+	ret = config->timing_source->stop(&config->timing_config, &data->timing_data);
+	if (ret != 0) {
+		LOG_ERR("Failed to stop timing source: %d", ret);
+		return ret;
+	}
+	stepper_trigger_callback(dev, STEPPER_EVENT_STOPPED);
+
+	return 0;
+}
+
+static int z_stepper_motion_control_set_reference_position(const struct device *dev,
+							   int32_t position)
+{
+	struct stepper_motion_control_data *data = dev->data;
+
+	K_SPINLOCK(&data->lock) {
+		data->actual_position = position;
+	}
+	return 0;
+}
+
+static int z_stepper_motion_control_get_actual_position(const struct device *dev, int32_t *position)
+{
+	struct stepper_motion_control_data *data = dev->data;
+
+	K_SPINLOCK(&data->lock) {
+		*position = data->actual_position;
+	}
+	return 0;
+}
+
+static int z_stepper_motion_control_set_step_interval(const struct device *dev,
+						      uint64_t microstep_interval_ns)
+{
+	const struct stepper_motion_control_config *config = dev->config;
+	struct stepper_motion_control_data *data = dev->data;
+
+	if (microstep_interval_ns == 0) {
+		LOG_ERR("Step interval is invalid.");
+		return -EINVAL;
+	}
+
+	K_SPINLOCK(&data->lock) {
+		data->timing_data.microstep_interval_ns = microstep_interval_ns;
+		config->timing_source->update(&config->timing_config, &data->timing_data,
+					      microstep_interval_ns);
+	}
+	LOG_DBG("Setting Motor step interval to %llu", microstep_interval_ns);
+	return 0;
+}
+
+static int z_stepper_motion_control_is_moving(const struct device *dev, bool *is_moving)
+{
+	const struct stepper_motion_control_config *config = dev->config;
+	struct stepper_motion_control_data *data = dev->data;
+
+	*is_moving = config->timing_source->is_running(&data->timing_data);
+	LOG_DBG("Motor is %s moving", *is_moving ? "" : "not");
+	return 0;
+}
+
+static void update_remaining_steps(const struct device *dev)
+{
+	struct stepper_motion_control_data *data = dev->data;
+
+	if (data->step_count > 0) {
+		data->step_count--;
+	} else if (data->step_count < 0) {
+		data->step_count++;
+	}
+}
+
+static void update_actual_position(const struct device *dev)
+{
+	struct stepper_motion_control_data *data = dev->data;
+
+	if (data->direction == STEPPER_DIRECTION_POSITIVE) {
+		data->actual_position++;
+	} else {
+		data->actual_position--;
+	}
```

Comment:
since we have only 2 enum values for direction (positive and negative), call of this function does not take into account if the stepper is moving at all or not. Calling it without a check would result in wrong position.

I would rather merge this function with `update_remaining_steps` or even execute the logic inline inside of both task functions because the actual logic is not really complex but extracting helper functions could introduce few ticks of execution overhead (function call overhead and data dereferencing).